sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+namespace Azure.Storage.Common.DataMovement
+{
+    /// <summary>
+    /// FilesystemScanner class
+    /// </summary>
+    public static class FilesystemScanner
+    {
+        /// <summary>
+        /// Enumerates all files pointed to by the provided path, including those in subdirectories (if path is a directory).
+        /// </summary>
+        /// <param name="path">Filesystem location.</param>
+        /// <returns>Enumerable list of absolute paths containing all relevant files the user has permission to access.</returns>
+        public static IEnumerable<string> ScanLocation(string path)
+        {
+            // Path type is ambiguous at start
+            bool isDirectory = false;
+
+            try
+            {
+                // Make sure we're dealing with absolute path
+                path = Path.GetFullPath(path);
+
+                // Check if path points to a directory
+                if ((File.GetAttributes(path) & FileAttributes.Directory) == FileAttributes.Directory)
+                {
+                    isDirectory = true;
+                }
+            }
+            catch
+            {
+                // If there's an error here, there aren't any valid entries to scan at the
+                // given path; either the path is invalid/nonexistant, or it isn't accessible by the user. In this case,
+                // don't return anything.
+                yield break;
```

Comment:
We should throw the exceptions that get thrown by `GetFullPath` instead of breaking and swallowing the exception.

Also we should add a TODO to log this error as well.

Reply:
Any specific logging technique to use, or should I just leave it to Console.WriteLine()?

Reply:
Definitely still add the TODO, but also we should throw any exceptions caught.

Reply:
Added the TODO, and set it to throw anything that comes up here.